ArduPlane/mode.cpp

```diff
@@ -0,0 +1,188 @@
+
+        // start with throttle suppressed in auto_throttle modes
+        plane.throttle_suppressed = plane.auto_throttle_mode;
+
+        plane.adsb.set_is_auto_mode(plane.auto_navigation_mode);
+    }
+
+    return enter_result;
+}
+
+bool Mode::has_manual_input() const
+{
+    return plane.g.stick_mixing != STICK_MIXING_DISABLED;
+}
+
+bool Mode::q_enter()
+{
+    plane.throttle_allows_nudging = true;
+    plane.auto_navigation_mode = false;
+    if (!plane.quadplane.init_mode() && plane.previous_mode != nullptr) {
+        plane.control_mode = plane.previous_mode;
+    } else {
+        plane.auto_throttle_mode = false;
+        plane.auto_state.vtol_mode = true;
+    }
+
+    return true;
+}
+
+void Mode::q_update()
+{
+    // set nav_roll and nav_pitch using sticks
+    int16_t roll_limit = MIN(plane.roll_limit_cd, plane.quadplane.aparm.angle_max);
+    plane.nav_roll_cd  = (plane.channel_roll->get_control_in() / 4500.0) * roll_limit;
+    plane.nav_roll_cd = constrain_int32(plane.nav_roll_cd, -roll_limit, roll_limit);
+    float pitch_input = plane.channel_pitch->norm_input();
+    // Scale from normalized input [-1,1] to centidegrees
+    if (plane.quadplane.tailsitter_active()) {
+        // For tailsitters, the pitch range is symmetrical: [-Q_ANGLE_MAX,Q_ANGLE_MAX]
+        plane.nav_pitch_cd = pitch_input * plane.quadplane.aparm.angle_max;
+    } else {
+        // pitch is further constrained by LIM_PITCH_MIN/MAX which may impose
+        // tighter (possibly asymmetrical) limits than Q_ANGLE_MAX
+        if (pitch_input > 0) {
+            plane.nav_pitch_cd = pitch_input * MIN(plane.aparm.pitch_limit_max_cd, plane.quadplane.aparm.angle_max);
+        } else {
+            plane.nav_pitch_cd = pitch_input * MIN(-plane.pitch_limit_min_cd, plane.quadplane.aparm.angle_max);
+        }
+        plane.nav_pitch_cd = constrain_int32(plane.nav_pitch_cd, plane.pitch_limit_min_cd, plane.aparm.pitch_limit_max_cd.get());
+    }
+}
+
+bool Mode::guided_enter()
+{
+    plane.throttle_allows_nudging = true;
+    plane.auto_throttle_mode = true;
+    plane.auto_navigation_mode = true;
+    plane.guided_throttle_passthru = false;
+    /*
+      when entering guided mode we set the target as the current
+      location. This matches the behaviour of the copter code
+    */
+    plane.guided_WP_loc = plane.current_loc;
+    plane.set_guided_WP();
+
+    return true;
+}
+
+void Mode::guided_update()
+{
+    if (plane.auto_state.vtol_loiter && plane.quadplane.available()) {
+        plane.quadplane.guided_update();
+    } else {
+        plane.calc_nav_roll();
+        plane.calc_nav_pitch();
+        plane.calc_throttle();
+    }
+}
+
+void Mode::fbwa_update()
+{
+    // set nav_roll and nav_pitch using sticks
+    plane.nav_roll_cd  = plane.channel_roll->norm_input() * plane.roll_limit_cd;
+    plane.nav_roll_cd = constrain_int32(plane.nav_roll_cd, -plane.roll_limit_cd, plane.roll_limit_cd);
```

Comment:
Probably preexsisting, but this line is a noop/waste of time. `RC_Channel::norm_input(void)` is internally (and logically) constrained to [-1, 1]